### ISSUE TITLE
Add `get_key` to `TabularMSA`

### ISSUE THREAD

**Issue by ebolyen** [biocore/scikit-bio#1125]
Regarding the interplay of `TabularMSA(seqs, key=KEY)` and other methods:

Users will generally fall into two camps when using the `TabularMSA`, they will either need keys, or they won't. Those that need keys will find it easiest to do something like `TabularMSA(seqs, key='id')` or equivalent. 

When interacting with other methods such as `append` there isn't a good reason why the msa couldn't recall the key used, preventing the user from having to duplicate the key for every interaction.

Here is a table indicating the success of two different `append` calls given different msa states: 

|                             |No key                    | `key='thing'`            | `keys=[...]`              |
|-----------------------------|--------------------------|--------------------------|---------------------------|
| `msa.append(seq)` | :heavy_check_mark:       | :white_check_mark:  |  :heavy_multiplication_x: |
| `msa.append(seq, key=key)`  | :heavy_multiplication_x: | :heavy_check_mark:       | :heavy_check_mark:        |

We want to add support for the green checkmark. Iit would be ideal if it succeeded as it is going to represent the most common call of a key-ed msa. 

If we intend to introduce operational state, we need a way for the user to access it. 

We propose adding a method `get_key` which would return the current key or raise an `OperationError` if it did not exist. This would also allow alignment functions to potentially propagate the key if two MSAs shared the same key, and would allow them to not necessarily require a key when aligning a novel sequence into an existing key-ed MSA. 

To support this change, `reindex` should set a private key variable, then other methods such as `append` can just call `get_key` if `has_keys` is `True` and a `key` was not provided.

**Comment by jairideout** [comment]
:+1:

**Comment by anderspitman** [comment]
:+1:

**Comment by anderspitman** [comment]
@ebolyen @jairideout what about calling the method `get_cached_key` or something similar, since `get_key` might be assumed to be a getter for the `key` property?

**Comment by jairideout** [comment]
If `TabularMSA` had a `key` property its getter would just be the `key` property itself (`msa.key`). Maybe this should be a `key` property instead of `get_key`?

**Comment by jairideout** [comment]
@ebolyen and I discussed this further. How about the following?

- `TabularMSA` constructor and `reindex` accept a `minter` parameter instead of a `key` parameter (this is just a name change). The minter describes how to create keys, so hopefully this will be more distinct from `keys`, which are keys that have already been created.
- `TabularMSA.minter` is a read-only property (i.e., just a getter) returning the minter or raising an `OperationError`. We could add a setter and deleter (allowing users to change how keys are created on an existing MSA) but this seems limited in its usefulness. We can always add this functionality later on without breaking backwards-compatibility if users ask for it.
- `TabularMSA.sort` remains the same, accepting a `key` parameter. This is to stay consistent with Python's sorting API, and this is actually a *sort key*, which is different from a minter since it only describes how to sort existing sequences.
- `TabularMSA.__eq__/__ne__` remain the same, such that two MSAs with identical data but different minters will compare equal. We can't reliably compare minters (e.g., an MSA with a lambda minter will never compare equal to another MSA with the same lambda minter) and minters describe how *future* keys will be created, so this is likely the behavior that most users will expect. Add a unit test for this behavior and document in docstring.

**Comment by ebolyen** [comment]
:+1:

**Comment by anderspitman** [comment]
Was there a particular reason to call the named parameter `minter` over `key_minter`?

**Comment by ebolyen** [comment]
Not really, `minter` is shorter, but `key_minter` may be clearer.

(sorry for closing, I always think I am commenting inline and want to close the box thing)

**Comment by wasade** [comment]
I like `minter`, but `key_forge` could be cool. Or `keyinator`

**Comment by anderspitman** [comment]
Yeah in the process of renaming "minter" seemed less precise. I'll go with "key_minter" if no one has any objections

**Comment by anderspitman** [comment]
Or maybe "keynter"?

**Comment by jairideout** [comment]
I like `minter` because it's shorter and more Pythonic. Also, the idea of a minter will appear on many of the central scikit-bio classes in the future so it may not be as big of an issue. What about keeping `minter` for now to see how users react to it? Since `TabularMSA` is experimental we can change the name if there's confusion.

That being said, `key_minter` is also a good name so no strong feelings.

**Comment by ebolyen** [comment]
Since it is experimental I think we should leave it `minter` and see how it works out. My general feeling is this is one of those things that in 2 years we would prefer to have made it just `minter`, but because it is the first time we have done this `key_minter` might seem better. 

There are several objects which will inevitably use this pattern (TreeNode, DistanceMatrix, All of the other MSAs, ContingencyTable, etc..) so in the long term, `minter` is probably a better name as it will appear very consistently, so any initial ambiguity will disappear after a few minutes with skbio.

**Comment by anderspitman** [comment]
`%s/key_minter/minter/g`

**Comment by jairideout** [comment]
@ebolyen and I chatted about storing the minter on `TabularMSA`. Storing the minter makes `TabularMSA` behavior more complex to reason about than we had expected, both to users and in implementation. For example:

- If a `TabularMSA` object has a stored `minter` and the user appends a sequence and provides a new minter, it is unclear what should happen. One option is to apply the minter to the appended sequence, but this creates an inconsistency because the stored minter may be different than the minter applied to the appended sequence. Thus, even though a stored minter exists, some of the keys on the MSA could come from a different source. Another option is to apply the new minter to *all* sequences (including the appended sequence) and store this new minter. This behavior (in addition to being inefficient) is likely not what the user expected: appending a sequence shouldn't reindex the entire MSA.

- Having a stored minter implies that keys must exist, but if keys exist that doesn't necessarily mean there is a minter. This isn't too terrible of a relationship between `keys` and `minter` but it makes for an awkward API (checking `has_keys` often isn't enough, and `has_minter` must also be called). I think this mostly makes implementing things like `append`, `extend`, `join`, etc. error-prone, but could also be confusing to users.

The primary motivation for having a minter is user-convenience, so we would like to keep a `minter` kwarg on the constructor, `reindex`, and `append` as this allows users to easily specify key generation. Storing this state, however, leads to complexity and doesn't really buy us much.

tl;dr: This issue can be closed by simply renaming `key` to `minter` :)

**Comment by ebolyen** [comment]
:boom:

**Comment by ebolyen** [comment]
This will make things a whole lot simpler.

**Comment by anderspitman** [comment]
Just to verify, the operating logic for append will now be:

Append call :point_down: MSA State :point_right:  | Has Keys | Doesn't Have Keys
---------|--------------|--------------------------
`msa.append()` | **FAIL** | **SUCCEED**
`msa.append(key='some string')` | **SUCCEED** unless duplicates | **FAIL**
`msa.append(minter='some string')` | **SUCCEED** unless not in metadata or duplicates | **FAIL**
`msa.append(minter=callable)` | **SUCCEED** unless bad minter logic or duplicates | **FAIL**

Does this look right?

**Comment by jairideout** [comment]
Nice table! Yep that all looks right.

**Comment by jairideout** [comment]
Fixed in #1121.
